packages/wdio-config/src/lib/ConfigParser.ts

```diff
@@ -26,9 +32,75 @@ interface MergeConfig extends Omit<Partial<TestrunnerOptionsWithParameters>, 'sp
     exclude?: string | string[]
 }
 
+// Get current working directory
+interface CurrentPathFinder {
+    getcwd(): string
+}
+
+// Require a .js/.json/dotfile config file
+interface LoadConfigFile {
+    loadFile<T>(path: string): T
+}
+
+// Detect if a file is present
+interface IsFileDetector {
+    isFile(path: string): boolean
+}
+
+// Glob to find file paths matching a pattern
+interface Globber {
+    glob(pattern: string): string[];
+}
+
+export interface PathService extends CurrentPathFinder, LoadConfigFile, IsFileDetector, Globber {}
+
+// Main implementation (tests contain other implementations)
+export class FileSystemPathService implements PathService {
+    getcwd(): string {
+        const cwd = process.cwd()
+        if ( typeof cwd === 'undefined' ) {
+            throw new Error('Unable to find current working directory from process')
+        }
+        return cwd
+    }
+
+    loadFile<T>(path: string): T {
+        if ( !path) {
+            throw new Error('A path is required')
+        }
+        return require(path)
+    }
+
+    isFile(filepath: string): boolean {
+        return (fs.existsSync(filepath) && fs.lstatSync(filepath).isFile())
+    }
+
+    glob(pattern: string): string[] {
+        return glob.sync(pattern)
+    }
+}
+
+class RequireLibrary implements ModuleRequireService {
+    require<T>(module: string): T {
+        return require(module) as T
+    }
+
+    resolve(request: string, options: { paths?: string[] }): string {
+        return require.resolve(request, options)
+    }
+}
```

Comment:
What is the purpose of those? If it is for testing purposes we could just use the mock features of jest, no? It seems they create a lot of overhead, especially when calling:

```js
exclude = ConfigParser.getFilePaths(capExclude, undefined, this._pathService)
```

Reply:
Yes they are for treating the file system as repositories. I desired testing different config files without needing to have more in the fixtures directory when all that changes is one property in them. In an earlier test I read the current fixture file into memory and modified it afterward, but it was feeling hacky.

As far as overhead I am not quite sure how that would be, it is passing a reference around to an object yes, but that object is running literally the same code as before. 

I think the mock features of jest would also work. I wanted to avoid coupling further with the implementation and mocking the process objects and felt capturing the abstract logic was cleaner. Both should absolutely work though, the big difference is where the hook for the tests happens: in application code as an interface instead of calling require, or in Node.js require infrastructure code providing mocked require responses for the code under test.